File: Src/ButterflyNet.Parser/HtmlAnalyzer.cs
using System;
using System.IO;
using System.Web;

namespace ButterflyNet.Parser {
	public class HtmlAnalyzer : ButterflyAnalyzer {

		public HtmlAnalyzer(TextWriter writer = null) : base(writer) { }

		public SyntaxHighlightingLibrary SyntaxHighlighter { get; set; }

		public override void OpenStrongText() {
			Writer.Write("<strong>");
		}

		public override void CloseStrongText() {
			Writer.Write("</strong>");
		}

		public override void OpenEmphasizedText() {
			Writer.Write("<em>");
		}

		public override void CloseEmphasizedText() {
			Writer.Write("</em>");
		}

		public override void OpenTeletypeText() {
			Writer.Write("<tt>");
		}

		public override void CloseTeletypeText() {
			Writer.Write("</tt>");
		}

		public override void OpenHeader(int depth) {
			depth = Math.Min(6, depth);
			Writer.Write("<h" + depth + ">");
		}

		public override void CloseHeader(int depth) {
			depth = Math.Min(6, depth);
			Writer.Write("</h" + depth + ">\n");
		}

		public override void WriteAndEscape(string text) {
			Writer.Write(HttpUtility.HtmlEncode(text));
		}

		public override void WriteAndEscape(char c) {
			Writer.Write(HttpUtility.HtmlEncode(c));
		}

		public override void WriteUnescapedString(string text) {
			Writer.Write(text);
		}

		public override void WriteUnescapedChar(char c) {
			Writer.Write(c);
		}

		public override void CloseParagraph() {
			Writer.Write("</p>\n");
		}

		public override void OpenUnderlinedText() {
			Writer.Write("<ins>");
		}

		public override void CloseUnderlinedText() {
			Writer.Write("</ins>");
		}

		public override void OpenStrikeThroughText() {
			Writer.Write("<del>");
		}

		public override void CloseStrikeThroughText() {
			Writer.Write("</del>");
		}

		public override void OpenHorizontalRuler() {
			Writer.Write("<hr />\n");
		}

		public override void OpenSmallText() {
			Writer.Write("<small>");
		}

		public override void CloseSmallText() {
			Writer.Write("</small>");
		}

		public override void OpenBigText() {
			Writer.Write("<big>");
		}

		public override void CloseBigText() {
			Writer.Write("</big>");
		}

		public override void OpenParagraph() {
			Writer.Write("<p>");
		}

		public override void OpenLink(string url, string baseUrl) {
			var cls = "class=\"external\" ";
			if (!url.IsExternalUrl()) {
				cls = "";
				url = baseUrl + url;
			}

			Writer.Write(string.Format("<a {1}href=\"{0}\">", HttpUtility.HtmlEncode(url), cls));
		}

		public override void CloseLink() {
			Writer.Write("</a>");
		}

		public override void OpenModule(IButterflyModule module) {
			module.Render(Writer);
		}

		public override void OpenBlockquote() {
			Writer.Write("<blockquote>");
		}

		public override void CloseBlockquote() {
			Writer.Write("</blockquote>\n");
		}

		public override void OpenOrderedList() {
			Writer.Write("<ol>\n");
		}

		public override void CloseOrderedList() {
			Writer.Write("</ol>\n");
		}

		public override void OpenUnorderedList() {
			Writer.Write("<ul>\n");
		}

		public override void CloseUnorderedList() {
			Writer.Write("</ul>\n");
		}

		public override void OpenListItem() {
			Writer.Write("<li>");
		}

		public override void CloseListItem() {
			Writer.Write("</li>\n");
		}

		public override void OpenPreformatted(string language) {
			var cls = string.Empty;
			if (!string.IsNullOrWhiteSpace(language)) {
				switch (SyntaxHighlighter) {
					case SyntaxHighlightingLibrary.Sunlight:
						cls = string.Format(" class=\"sunlight-highlight-{0}\"", language);
						break;
					case SyntaxHighlightingLibrary.SyntaxHighlighter:
						cls = string.Format(" class=\"brush: {0}\"", language);
						break;
					case SyntaxHighlightingLibrary.Prettify:
						cls = " class=\"prettyprint\"";
						break;
					case SyntaxHighlightingLibrary.HighlightJs:
						cls = string.Format(" class=\"language-{0}\"", language);
						break;
				}
			}

			Writer.Write(string.Format("<pre{0}>", cls));
		}

		public override void ClosePreformatted(string language) {
			Writer.Write("</pre>\n");
		}

		public override void OpenPreformattedLine() {
			Writer.Write("<pre>");
		}

		public override void ClosePreformattedLine() {
			ClosePreformatted(null);
		}

		public override void OpenTable() {
			Writer.Write("<table>\n");
		}

		public override void CloseTable() {
			Writer.Write("</table>\n");
		}

		public override void OpenTableRow() {
			Writer.Write("<tr>\n");
		}

		public override void CloseTableRow() {
			Writer.Write("</tr>\n");
		}

		public override void OpenTableRowLine() {
			OpenTableRow();
		}

		public override void CloseTableRowLine() {
			CloseTableRow();
		}

		public override void OpenTableHeader() {
			Writer.Write("<th>");
		}

		public override void CloseTableHeader() {
			Writer.Write("</th>\n");
		}

		public override void OpenTableCell() {
			Writer.Write("<td>");
		}

		public override void CloseTableCell() {
			Writer.Write("</td>\n");
		}

		public override void OpenDefinitionList() {
			Writer.Write("<dl>");
		}

		public override void CloseDefinitionList() {
			Writer.Write("</dl>\n");
		}

		public override void OpenDefinitionTerm() {
			Writer.Write("<dt>");
		}

		public override void CloseDefinitionTerm() {
			Writer.Write("</dt>\n");
		}

		public override void OpenDefinition() {
			Writer.Write("<dd>");
		}

		public override void CloseDefinition() {
			Writer.Write("</dd>\n");
		}

		public override void CloseMultiLineDefinition() {
			CloseDefinition();
		}

		public override void OpenMultiLineDefinition() {
			OpenDefinition();
		}

		public override void OpenLineBreak() {
			Writer.Write("<br />");
		}

	}
}

File: src/js/src/HtmlAnalyzer.js
var SyntaxHighlightingLibrary = {
	sunlight: "sunlight",
	syntaxHighlighter: "SyntaxHighlighter",
	prettify: "prettify",
	highlightJs: "highlightjs",
	none: "none"
};

function htmlEncode(text) {
	return text.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/'/g, "&#39;");
}

function isExternalUrl(url) {
	return /^[^\/]:/.test(url || "");
}

function HtmlAnalyzer(writer) {
	HtmlAnalyzer.$parent.call(this, writer);
	
	function convertToString(value) {
		if (typeof(text) === "number") {
			text = String.fromCharCode(text);
		}
		
		return text;
	}
	
	this.syntaxHighlighter = SyntaxHighlightingLibrary.sunlight;
	
	this.openStrongText = function() { this.writer.write("<strong>"); };
	this.closeStrongText = function() { this.writer.write("</strong>"); };
	
	this.openEmphasizedText = function() { this.writer.write("<em>"); };
	this.closeEmphasizedText = function() { this.writer.write("</em>"); };
	
	this.openTeletypeText = function() { this.writer.write("<tt>"); };
	this.closeTeletypeText = function() { this.writer.write("</tt>"); };
	
	this.openHeader = function(depth) { this.writer.write("<h" + Math.min(6, depth) + ">"); };
	this.closeHeader = function(depth) { this.writer.write("</h" + Math.min(6, depth) + ">\n"); };
	
	this.writeAndEscape = function(text) { this.writer.write(htmlEncode(convertToString(text))); };
	this.writeUnescapedString = function(text) { this.writer.write(convertToString(text)); };
	
	this.openParagraph = function() { this.writer.write("<p>"); };
	this.closeParagraph = function() { this.writer.write("</p>\n"); };
	
	this.openUnderlinedText = function() { this.writer.write("<ins>"); };
	this.closeUnderlinedText = function() { this.writer.write("</ins>"); };
	
	this.openStrikeThroughText = function() { this.writer.write("<del>"); };
	this.closeStrikeThroughText = function() { this.writer.write("</del>"); };
	
	this.openHorizontalRuler = function() { this.writer.write("<hr />\n"); };
	
	this.openSmallText = function() { this.writer.write("<small>"); };
	this.closeSmallText = function() { this.writer.write("</small>"); };
	
	this.openBigText = function() { this.writer.write("<big>"); };
	this.closeBigText = function() { this.writer.write("</big>"); };
	
	this.openLink = function(url, baseUrl) {
		var cls = "class=\"external\" ";
		if (!isExternalUrl(url)) {
			cls = "";
			url = baseUrl + url;
		}
		
		this.writer.write("<a " + cls + "href=\"" + htmlEncode(url) + "\">");
	};
	this.closeLink = function() { this.writer.write("</a>"); };
	
	this.openModule = function(module) { module.render(buffer); };
	
	this.openBlockquote = function() { this.writer.write("<blockquote>"); };
	this.closeBlockquote = function() { this.writer.write("</blockquote>\n"); };
	
	this.openOrderedList = function() { this.writer.write("<ol>\n"); };
	this.closeOrderedList = function() { this.writer.write("</ol>\n"); };
	
	this.openUnorderedList = function() { this.writer.write("<ul>\n"); };
	this.closeUnorderedList = function() { this.writer.write("</ul>\n"); };
	
	this.openListItem = function() { this.writer.write("<li>"); };
	this.closeListItem = function() { this.writer.write("</li>\n"); };
	
	this.openPreformatted = function(language) {
		var cls = "",
			className;
		if (language) {
			switch (this.syntaxHighlighter) {
				case SyntaxHighlightingLibrary.sunlight:
					className = "sunlight-highlight-" + language;
					break;
				case SyntaxHighlightingLibrary.syntaxHighlighter:
					className = "brush: " + language;
					break;
				case SyntaxHighlightingLibrary.prettify:
					className = "prettyprint";
					break;
				case SyntaxHighlightingLibrary.highlightJs:
					className = "language-" + language;
					break;
			}
			
			cls = " class=\"" + className + "\"";
		}
		
		this.writer.write("<pre" + cls + ">");
	};
	
	this.closePreformatted = function(language) { this.writer.write("</pre>\n"); };
	
	this.openPreformattedLine = function(language) { this.writer.write("<pre>"); };
	this.closePreformattedLine = this.closePreformatted;
	
	this.openTable = function() { this.writer.write("<table>\n"); };
	this.closeTable = function() { this.writer.write("</table>\n"); };
	
	this.openTableRow = function() { this.writer.write("<tr>\n"); };
	this.closeTableRow = function() { this.writer.write("</tr>\n"); };
	
	this.openTableRowLine = this.openTableRow;
	this.closeTableRowLine = this.closeTableRow;
	
	this.openTableHeader = function() { this.writer.write("<th>"); };
	this.closeTableHeader = function() { this.writer.write("</th>\n"); };
	
	this.openTableCell = function() { this.writer.write("<td>"); };
	this.closeTableCell = function() { this.writer.write("</td>\n"); };
	
	this.openDefinitionList = function() { this.writer.write("<dl>\n"); };
	this.closeDefinitionList = function() { this.writer.write("</dl>\n"); };
	
	this.openDefinitionTerm = function() { this.writer.write("<dt>"); };
	this.closeDefinitionTerm = function() { this.writer.write("</dt>\n"); };
	
	this.openDefinition = function() { this.writer.write("<dd>"); };
	this.closeDefinition = function() { this.writer.write("</dd>\n"); };
	
	this.closeMultiLineDefinition = this.openDefinition;
	this.closeMultiLineDefinition = this.closeDefinition;
	
	this.openLineBreak = function() { this.writer.write("<br />"); };
}

extend(ButterflyAnalyzer, HtmlAnalyzer);


